Source/UnitTests/Common/MathUtilTest.cpp

```diff
@@ -26,3 +26,39 @@ TEST(MathUtil, NextPowerOf2)
   EXPECT_EQ(8U, MathUtil::NextPowerOf2(6));
   EXPECT_EQ(0x40000000U, MathUtil::NextPowerOf2(0x23456789));
 }
+
+TEST(MathUtil, SaturatingCast)
+{
+  // Cast from an integer type to a smaller type
+  EXPECT_EQ(255u, (MathUtil::SaturatingCast<u8, int>(1000)));
+  EXPECT_EQ(255u, (MathUtil::SaturatingCast<u8, u16>(1000u)));
+  EXPECT_EQ(255u, (MathUtil::SaturatingCast<u8, std::size_t>(1000)));
+
+  // Cast from a signed integer type
+  EXPECT_EQ(0u, (MathUtil::SaturatingCast<u8, int>(-1)));
+  EXPECT_EQ(0u, (MathUtil::SaturatingCast<u8, int>(-1000)));
+  EXPECT_EQ(0u, (MathUtil::SaturatingCast<u32, int>(-1)));
+  EXPECT_EQ(-1000, (MathUtil::SaturatingCast<s16, int>(-1000)));
+  EXPECT_EQ(-1000, (MathUtil::SaturatingCast<int, int>(-1000)));
+  EXPECT_EQ(-1000, (MathUtil::SaturatingCast<s64, int>(-1000)));
+
+  // Cast from an unsigned integer type to a smaller integer type
+  EXPECT_EQ(0x7fff, (MathUtil::SaturatingCast<s16, u32>(0xffffffffu)));
+  EXPECT_EQ(0x7fffffff, (MathUtil::SaturatingCast<int, u32>(0xffffffffu)));
+
+  // Cast from a floating point type to an integer type
+  EXPECT_EQ(255u, MathUtil::SaturatingCast<u8>(1234.0));
+  EXPECT_EQ(0u, MathUtil::SaturatingCast<u8>(-1234.0));
+  EXPECT_EQ(127, MathUtil::SaturatingCast<s8>(5678.0));
+  EXPECT_EQ(-128, MathUtil::SaturatingCast<s8>(-5678.0));
+  EXPECT_EQ(65535u, MathUtil::SaturatingCast<u16>(999999.0));
+
+  // Edge cases
+  EXPECT_EQ(std::numeric_limits<s32>::max(),
+            MathUtil::SaturatingCast<s32>(std::numeric_limits<float>::infinity()));
+  EXPECT_EQ(std::numeric_limits<s32>::min(),
+            MathUtil::SaturatingCast<s32>(-std::numeric_limits<float>::infinity()));
```

Comment:
What about NaN?  (Not sure if there's a reasonable way to handle NaN).

It might also be worth adding a test case for -0, though -0 is already covered in VertexLoaderTest.cpp.  Say, `EXPECT_EQ(0u, MathUtil::SaturatingCast<u8>(-0.0));` and `EXPECT_EQ(0, MathUtil::SaturatingCast<s8>(-0.0));`

Reply:
Yeah, I'm unsure how NaN should be treated. Tempted to just leave it undefined, because it's unlikely to come up in practice and it doesn't make sense to return either lo or hi.

I'll add tests for -0.0, good idea.

Reply:
A comment saying "the result of SaturatingCast with a value of NaN is undefined" would be fine by me.